### ISSUE TITLE
Prism launcher can't utilize the dedicated GPU

### ISSUE THREAD

**Issue by 51k34** [PrismLauncher/PrismLauncher#1035]
### Operating System

Linux

### Version of Prism Launcher

Prism Launcher 6.3

### Version of Qt

Qt 5.15.9

### Description of bug

Prism launcher does not utilize the discrete (AMD) GPU, instead using the integrated (AMD)GPU even when setting use discrete GPU. i have also:

Put my PC in discrete GPU only mode from my OEM's hardware utilization software, and continued to have the setting on through the other experiments

I have used the official arch Minecraft launcher, which did use the discrete GPU.

Also played with the exact same mods and settings on windows, which DID use the dedicated gpu

Launched the launcher from Lutris and setting Lutris to also only use the discrete GPU in the settings(this usually works).

Set Prism to use feral gamemode and all native libraries. 

I've also checked Flatseal for a setting to use the GPU.

Edited the .sh script in `/var/lib/flatpak/app/org.prismlauncher.PrismLauncher/current/active/export/bin/` to launch Prism with the default or none default GPU. using the command `PrefersNonDefaultGPU=true` and `PrefersNonDefaultGPU=false`

inxi:
```
System:
  Host: simon-m-zephyrusg14-ga402rkga402rk Kernel: 6.1.25-1-MANJARO
    arch: x86_64 bits: 64 Desktop: KDE Plasma v: 5.27.4 Distro: Manjaro Linux

Machine:
  Type: Laptop System: ASUSTeK product: ROG Zephyrus G14 GA402RK_GA402RK
    v: 1.0 serial: <superuser required>
  Mobo: ASUSTeK model: GA402RK v: 1.0 serial: <superuser required>
    UEFI: American Megatrends LLC. v: GA402RK.318 date: 03/09/2023

CPU:
  Info: 8-core model: AMD Ryzen 9 6900HS with Radeon Graphics bits: 64
    type: MT MCP cache: L2: 4 MiB
  Speed (MHz): avg: 3061 min/max: 1600/4934 cores: 1: 3300 2: 3300 3: 1391
    4: 3300 5: 3300 6: 3300 7: 3300 8: 3300 9: 3300 10: 3300 11: 1392 12: 3300
    13: 3300 14: 3300 15: 3300 16: 3300

Graphics:
  Device-1: AMD Navi 23 [Radeon RX 6650 XT / 6700S 6800S] driver: amdgpu
    v: kernel
  Device-2: AMD Rembrandt [Radeon 680M] driver: amdgpu v: kernel
  Device-3: Quanta USB2.0 HD UVC WebCam type: USB driver: uvcvideo
  Display: wayland server: X.org v: 1.21.1.8 with: Xwayland v: 23.1.1
    compositor: kwin_wayland driver: X: loaded: amdgpu
    unloaded: modesetting,radeon dri: radeonsi gpu: amdgpu,amdgpu
    resolution: 1707x1067
  API: OpenGL v: 4.6 Mesa 23.0.2 renderer: AMD Radeon RX 6800S (navi23 LLVM
    15.0.7 DRM 3.49 6.1.25-1-MANJARO)

Audio:
  Device-1: AMD Navi 21/23 HDMI/DP Audio driver: snd_hda_intel
  Device-2: AMD Rembrandt Radeon High Definition Audio driver: snd_hda_intel
  Device-3: AMD ACP/ACP3X/ACP6x Audio Coprocessor driver: snd_pci_acp6x
  Device-4: AMD Family 17h/19h HD Audio driver: snd_hda_intel
  API: ALSA v: k6.1.25-1-MANJARO status: kernel-api
  Server-1: PulseAudio v: 16.1 status: active

Drives:
  Local Storage: total: 1.86 TiB used: 157.06 GiB (8.2%)
  ID-1: /dev/nvme0n1 vendor: Toshiba model: N/A size: 1.86 TiB

Partition:
  ID-1: / size: 891.53 GiB used: 77.6 GiB (8.7%) fs: ext4 dev: /dev/nvme0n1p5
  ID-2: /boot/efi size: 256 MiB used: 67.8 MiB (26.5%) fs: vfat
    dev: /dev/nvme0n1p1
  ID-3: /home size: 93.13 GiB used: 79.4 GiB (85.2%) fs: ntfs
    dev: /dev/nvme0n1p6

Swap:
  Alert: No swap data was found.

Sensors:
  System Temperatures: cpu: 62.2 C mobo: N/A
  Fan Speeds (RPM): cpu: 2000
  GPU: device: amdgpu temp: 57.0 C fan: 0 device: amdgpu temp: 56.0 C

Info:
  Processes: 380 Uptime: 22m Memory: 30.59 GiB used: 6.41 GiB (20.9%)
  Shell: Zsh inxi: 3.3.26
```

### Steps to reproduce

- Launch Prism Launcher on Linux

Steps to check whether or not Prism Launcher is using the dedicated GPU:
- use nvtop
- see if the integrated or dedicated GPU is using more resources in a system monitoring app

### Suspected cause

Prism launcher not detecting the GPU

yeah that's pretty much all i could think of being the case, I've tried everything else

### This issue is unique

- [x] I have searched the issue tracker and did not find an issue describing my bug.

**Comment by 51k34** [comment]
I have also tested the following versions:

- the appimage version of prism
- the tar.gz version of prism
- the .bin from the arch user repository

and still with no success

**Comment by Jan200101** [comment]
`PrefersNonDefaultGPU` is not advised as its behavior is poorly implemented across the board.

Prism behaves the same as Lutris and sets the right environment variables
https://github.com/PrismLauncher/PrismLauncher/blob/ce5bb29c442cee3654c5f4287a999d5d6593032f/launcher/minecraft/MinecraftInstance.cpp#L520-L524

I've tested this on Prism 6.3 on my Laptop with a discrete Nvidia RTX GPU, and it works without a problem.

**Comment by 51k34** [comment]
> `PrefersNonDefaultGPU` is not advised as its behavior is poorly implemented across the board.

i was aware, but as i was trying many different methods and options to try and solve this issue i did it anyway.

**Comment by body20002** [comment]
I have the same issue, I use prism flatpak
what's weird is when I set the global settings to use `discrete GPU` the game do not launch and crash and if I only changed the `instance` settings the game launches but not with `discrete GPU`

here are the logs https://pastebin.com/XjiKQN4i

**Comment by cfazilleau** [comment]
Same issue here

**Comment by lichenaut** [comment]
What worked for me when I had a similar issue was creating my own prime-run:

`alias prime-run='__NV_PRIME_RENDER_OFFLOAD=1 __GLX_VENDOR_LIBRARY_NAME=nvidia'`

then:

`prime-run /usr/bin/flatpak run --branch=stable --arch=x86_64 --command=prismlauncher org.prismlauncher.PrismLauncher`


EDIT: now I just 'prime-run prismlauncher'

**Comment by ghost** [comment]
Or you can add javaw.exe to (Windows > Setting > Gaming > Game Mode > Graphics) and set High performance preference to it

**Comment by SpiderUnderUrBed** [comment]
![image](https://github.com/PrismLauncher/PrismLauncher/assets/71126384/72aa24f6-1ca9-42dc-8a28-7c9788babef3)

^ this worked for me

**Comment by Lyapsus** [comment]
There is simply no “use dedicated gpu” option

**Comment by mm12** [comment]
> Or you can add javaw.exe to (Windows > Setting > Gaming > Game Mode > Graphics) and set High performance preference to it


This is a band-aid solution that works. Adding to the AMD software does not seem to work. 


> [image]
> ^ this worked for me

There is no "performance" tab at all on my launcher - I am on Windows

**Comment by binarydigitz01** [comment]
> What worked for me when I had a similar issue was creating my own prime-run:
> 
> `alias prime-run='__NV_PRIME_RENDER_OFFLOAD=1 __GLX_VENDOR_LIBRARY_NAME=nvidia'`
> 
> then:
> 
> `prime-run /usr/bin/flatpak run --branch=stable --arch=x86_64 --command=prismlauncher org.prismlauncher.PrismLauncher`
> 
> EDIT: now I just 'prime-run prismlauncher'

I have the same Problem, and this worked for me.

**Comment by S-h-i-n-r-a** [comment]
> `PrefersNonDefaultGPU` is not advised as its behavior is poorly implemented across the board.
> 
> Prism behaves the same as Lutris and sets the right environment variables
> 
> https://github.com/PrismLauncher/PrismLauncher/blob/ce5bb29c442cee3654c5f4287a999d5d6593032f/launcher/minecraft/MinecraftInstance.cpp#L520-L524
> 
> I've tested this on Prism 6.3 on my Laptop with a discrete Nvidia RTX GPU, and it works without a problem.

I notice you're setting the DRI_PRIME env var to 1, this in the case of my system is the integrated graphics with the dedicated GPU being 0, presumably this - in my case - is caused by switching from a CPU without integrated graphics to one that does have it.
This may also be the case for OP.

A solution inf this is the case would be to:
- find or provide a way to change this setting
- change the env var after prism does
- for OP (and me) to compile prismlauncher with that value changed in the cpp file
- for OP and me to somehow change the indices of the GPUs

![image](https://github.com/user-attachments/assets/5303edaf-ca02-4ad1-9c20-c1808a7ffde9)

(EDIT: looking at OP's inxi this is a me issue)

**Comment by Trial97** [comment]
you can already set any env variables you want inside the Environment variables tab inside either instance settings( see photo) or the main settings
![image](https://github.com/user-attachments/assets/d392e1f9-62fd-4de3-a7a0-3e580e822f3f)

So for your case just copy the env from the file and change the values in that setting tab. The `Use dedicated GPU` turn it off

**Comment by S-h-i-n-r-a** [comment]
After some testing neither DRI_PRIME=0 nor DRI_PRIME=1 ended up running the game with discrete graphics*, I disabled integrated graphics in the BIOS to resolve the issue.

*: F3 menu showed the integrated graphics in both cases

**Comment by Jan200101** [comment]
> I notice you're setting the DRI_PRIME env var to 1, this in the case of my system is the integrated graphics with the dedicated GPU being 0, presumably this - in my case - is caused by switching from a CPU without integrated graphics to one that does have it.

the issue is that Mesa is naive and DRI_PRIME=1 means nothing more than "use the second GPU" regardless of that GPU is actually discrete or not.

I'm currently working on changing the implementation to use switcheroo-control instead and expose a list of all available GPUs to choose from.

**Comment by owendor** [comment]
> I'm currently working on changing the implementation to use switcheroo-control instead and expose a list of all available GPUs to choose from.

Any progress on this?

**Comment by Jan200101** [comment]
> Any progress on this?

Currently waiting on my current GPU PR to get reviewed and merged

**Comment by andrpetim** [comment]
Alright, maybe i`m not on the theme and i`m windows user, but can i do something to enable Perfomance tab in setting? I just dont have it and i think minecraft using integrated Intel card, though i have Nvidia card
I think so because Nvidium mod doesnt have any effect visually

**Comment by Trial97** [comment]
@andrpetim the Performance Tab is useless on Windows. 
Please refer to https://github.com/PrismLauncher/PrismLauncher/issues/2832

**Comment by RainOfPain125** [comment]
gonna chime in and report the same kind of thing instead of making a new issue.

Linux Mint 22
Prism Launcher 9.1
Qt 6.8.1
Not using dedicated GPU

I have an LG 17U70Q laptop. Prism defaults to using Intel Iris Graphics instead of the RTX 3050TI even with "Use discrete GPU" tweak enabled.

I've reviewed some of the "solutions" given in this thread, but one of the easiest - "disable onboard gpu in BIOS" is both not available in my BIOS, and would not be very optimized for battery life.

Other solutions I skimmed and seemed more complicated. I might circle back to try them and post results.

update: lets just say I'm dumb. secure boot enabled, Nvidia drivers never loaded. to be fair, I think if you have Linux and you chose to install the proprietary driver, there should be a gigantic warning with every login saying "YOUR GPU DRIVERS ARE INSTALLED, BUT DID NOT LOAD BECAUSE THEY ARE NOT VERIFIED IN SECURE BOOT" but instead there is no warning.

very sorry! No fault on Prism Launcher, Minecraft now opens with the dedicated GPU as expected.

**Comment by AoSankaku** [comment]
> Or you can add javaw.exe to (Windows > Setting > Gaming > Game Mode > Graphics) and set High performance preference to it

This issue shouldn't be stale, this is an ongoing issue with a significant performance bottlenecking.

And I suppose the native and best way to optimize performance is to do this. Isn't it possible to do this automatically by implementing something to PrismLauncher to solve this issue?

**Comment by AoSankaku** [comment]
https://github.com/PrismLauncher/PrismLauncher/issues/444#issuecomment-2730625194

This doesn't seem to be a valid option for Prism v9.4

**Comment by piscopancer** [comment]
hello, my gpu is not seen by minecraft either, 2025

**Comment by TheRedScreen64** [comment]
Fix for people using KDE Plasma (I'm using fedora): Search for Prism Launcher in Startmenu, righ click, edit application, go to advanced tab, check run using dedicated graphics card.

If you don't see that option, make sure that the rpmfusion nvidia driver is installed correctly (when you have an nvidia gpu as dedicated).

**Comment by Keffori** [comment]
Yesterday everything worked perfectly with "prime-run" typed everyone in "Custom Commands" window, today it just broke. I didn't update the system. No problems found on other launchers, so I think its a problem with Prism itself

**Comment by luckyone89** [comment]
I have started having this exact problem recently. I tried to open the application using my discrete GPU. I also tried to place DRI_PRIME=1 or 0 in the environment variables or launch commands anywhere I could find, but nothing worked. Launching another application using any technique that I describe works, so I also believe it is something within the application that breaks this functionality.

**Comment by Handrail9** [comment]
Seeing this issue on AMD Framework 16 with AMD dGPU running Fedora 42 GNOME.

**Comment by Sloth-on-meth** [comment]
same problem here

**Comment by andreea-mgh** [comment]
I had this problem on linux flatpak. The "discrete GPU" setting in Performance settings didn't work. `prime-run` also didn't work. They all resulted in the launcher (and then the Minecraft instance) both falling back to integrated graphics.

The issue was that the Prism Launcher FLATPAK is not configured to have access to the GPU. You can manually enable this using Flatseal. In Flatseal the permission "GPU acceleration" (or `device=dri`) must be turned on. I don't understand how it's not enabled by default for a game...

**Comment by gabetronx** [comment]
> I had this problem on linux flatpak. The "discrete GPU" setting in Performance settings didn't work. `prime-run` also didn't work. They all resulted in the launcher (and then the Minecraft instance) both falling back to integrated graphics.
> 
> The issue was that the Prism Launcher FLATPAK is not configured to have access to the GPU. You can manually enable this using Flatseal. In Flatseal the permission "GPU acceleration" (or `device=dri`) must be turned on. I don't understand how it's not enabled by default for a game...

This worked for me. Thank you. AMD CPU and AMD GPU, and this finally worked for me.

**Comment by cypressSap** [comment]
Same issue on Debian 12 after working perfectly fine for over a year. Was still working last time I played, which according to the logs for my usual instance was June 29.

Fresh vanilla 26.1.2 instance in the official launcher (60 FPS):
> AMD Radeon R9 200 Series (tonga, LLVM 15.0.6, DRM 3.49, 6.1.0-50-amd64)
> OpenGL 4.6 (Core Profile) Mesa 22.3.6

Fresh vanilla 26.1.2 instance in Prism (1-5 FPS):
> llvmpipe (LLVM 21.1.8, 256 bits)
> OpenGL 4.6 (Core Profile) Mesa 26.1.4 (git-6dfbc555b4)

Enabling `device=dri` permission and/or checking "Use discrete GPU" in Prism have not helped.

Rolling back the kernel and/or flatpak image have also not helped.

Edit: This is specifically an issue with the flatpak version; the "portable" version works fine.
